### ISSUE TITLE
[FEATURE REQUEST] Custom filters in SpladeTable Component 

### ISSUE THREAD

**Issue by luisvbz** [protonemedia/laravel-splade#359]
- Laravel Version: 10.3.3
- PHP Version: 8.1.17
- Splade JS Version (npm): 1.2.2
- Splade PHP Version (composer): 1.4.1

Hello, it would be great to have the possibility of adding custom filters in SpladeTable tables with a slot and thus adding these filters with the Splade form components.
For example: 

```html
<x-splade-table :for="$users">
    <x-slot name="filters">
           <!-- The key attribute would be the column by which it would filter -->
           <x-splade-input label="Date of Creation" name="created_at"  key="created_at" date range />
           <x-splade-select label="Country" name="country" :options="$countires" key="country_id" choices />
    </x-slot>
</x-splade-table>
```

So we could use these filters automatically by the **key** attributte or using the **When** function within the SpladeTable class
For exmaple:

```php
class Users extends AbstractTable
{

    public function for()
    {
        return User::query()->when($this->filters['created_at'], function($q) {
                         $q->whereBetween('created_at', $this->filters['created_at'])
                   })->when($this->filters['country'], fn($q) => $q->where('country_id', $this->filters['country']);
    }
}
```
By the way, excellent work!

**Comment by luisvbz** [comment]
Yes, I know it has those filters. Actually, I use them quite often.
My idea or proposal is to enable the "filters" slot to allow you to add any filter you need, such as an input with a date range, filters with checkboxes, a multiple select, etc. This would give you greater control over what you want to search for, instead of just relying on the filters that the SpladeTable class currently has.

@zulfikar-ditya thanks for reply me

> Did you mean splade data table filter ? Did you look this doc. I don't now this help or not but [here](https://splade.dev/docs/table-spatie-query-builder) Try using [splade table](https://splade.dev/docs/table-overview) in Table Class section
> 
> ```shell
> php artisan make:table Users
> ```
> 
> will look like this
> 
> ```
> $table
> // ...other column's 
> ->selectFilter(
>     key: 'id', 
>     label: 'Country',
>     options: \App\Models\Country::all(['id', 'name'])->toArray()
> )
> // ...other column's 
> ```
> 
> @luisvbz CMIIW

Yes, I know it has those filters. Actually, I use them quite often.
My idea or proposal is to enable the "filters" slot to allow you to add any filter you need, such as an input with a date range, filters with checkboxes, a multiple select, etc. This would give you greater control over what you want to search for, instead of just relying on the filters that the SpladeTable class currently has.

**Comment by zulfikar-ditya** [comment]
Ah my bad, 
I don't recognize that, but date range picker is in progress #162 by @pascalbaljet . Maybe we would like to filter with multiple select value to or filter using checkbox like you said @luisvbz

**Comment by luisvbz** [comment]
Any updates?

**Comment by zulfikar-ditya** [comment]
#162 stil in draft

**Comment by e-smily** [comment]
> Did you mean splade data table filter ? Did you look this doc. I don't now this help or not but [here](https://splade.dev/docs/table-spatie-query-builder) Try using [splade table](https://splade.dev/docs/table-overview) in Table Class section
> 
> ```shell
> php artisan make:table Users
> ```
> 
> will look like this
> 
> ```
> $table
> // ...other column's 
> ->selectFilter(
>     key: 'id', 
>     label: 'Country',
>     options: \App\Models\Country::all(['id', 'name'])->toArray()
> )
> // ...other column's 
> ```
> 
> @luisvbz CMIIW

Always get:

`htmlspecialchars(): Argument #1 ($string) must be of type string, array given`

when i try to fill the options of the select-filter:

options: ObjektGruppe::all(['id', 'gruppe_name_de'])->toArray() ,

it only works with:

options: [8 => 'Gemälde', 22 => 'Porzellan',],

how can i dynamicly fill the select-filter?

**Comment by dusp2k** [comment]
> > Did you mean splade data table filter ? Did you look this doc. I don't now this help or not but [here](https://splade.dev/docs/table-spatie-query-builder) Try using [splade table](https://splade.dev/docs/table-overview) in Table Class section
> > ```shell
> > php artisan make:table Users
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will look like this
> > ```
> > $table
> > // ...other column's 
> > ->selectFilter(
> >     key: 'id', 
> >     label: 'Country',
> >     options: \App\Models\Country::all(['id', 'name'])->toArray()
> > )
> > // ...other column's 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @luisvbz CMIIW
> 
> Always get:
> 
> `htmlspecialchars(): Argument #1 ($string) must be of type string, array given`
> 
> when i try to fill the options of the select-filter:
> 
> options: ObjektGruppe::all(['id', 'gruppe_name_de'])->toArray() ,
> 
> it only works with:
> 
> options: [8 => 'Gemälde', 22 => 'Porzellan',],
> 
> how can i dynamicly fill the select-filter?

Use mapWithKeys() to flatten the array to key/value pairs:

options: ObjektGruppe::all()->mapWithKeys(fn($objektGruppe) =>[$objektGruppe->id => $objektGruppe->gruppe_name_de])

**Comment by e-smily** [comment]
Thank you Daniel, that helped.
For the sake of completeness, here is what it looks like now:

```
options: ObjektGruppe::all()->where('aktiv', '=', 1)->sortBy('gruppe_name_de')
                ->mapWithKeys(fn($objektGruppe) =>[$objektGruppe->id => $objektGruppe->gruppe_name_de])
                ->toArray(),
```

**Comment by e-smily** [comment]
I just found another problem, but maybe I'm just making a mistake?

The SQL string that is created:

`select count(*) as aggregate from `object` left join `object_group` on `object`.`object_group_id` = `object_group`.`id` and `object_group`.`deleted_at` is null where LOWER(`object`.` object_group_id`) LIKE '%2%' and `object`.`deleted_at` is null`

The problem is that the key is filtered as a string with LIKE:

`where LOWER(`object`.`object_group_id`) LIKE '%2%'`

So I don't get a unique group, it would have to be filtered for an integer:

`where `object`.`object_group_id` = 2`

how can I influence that?

**Comment by dusp2k** [comment]
Don't know about the LIKE, but this:

```
ObjektGruppe::all()->where('aktiv', '=', 1)->sortBy('gruppe_name_de')
```

This will first load all records from the DB and than do the where() and sortBy() on the collection.

You might want to do it the following way, which will do the filtering and ordering on the DB side:

```
ObjektGruppe::where('aktiv', '=', 1)->orderBy('gruppe_name_de', 'asc')->get()
```

**Comment by e-smily** [comment]
Unfortunately my English isn't the best :)

the select component is filled and displayed correctly, but when I try to filter and select an element, the SQL string described above is generated internally by Splade.. I have no influence on that..
For example, I have the following data in the "object groups" table (id is integer, group_name_de is text)

1 painting
2 glass
...
15 porcelain

If you now filter by `object_group_id LIKE '%1%'`, paintings and porcelain will be included in the selection
if `object_group_id = 1` is filtered, only paintings will be selected

**Comment by kahfieidn** [comment]
> ```
> filters
> ```
When i adding slot in table splade name="filters", in views showing nothing for button filter, any solve?

**Comment by kahfieidn** [comment]
See this #539 => https://github.com/protonemedia/laravel-splade/issues/539 

working fine for me, For now my only workaround is to make custom month & year filter, using spatie query builder
